lib/src/subscription.dart

```diff
@@ -0,0 +1,284 @@
+  error,
+
+  /// The server responded to a later subscription set to this one and this one has been
+  /// trimmed from the local storage of subscription sets.
+  superseded,
+}
+
+/// A collection representing the set of active subscriptions for a [Realm] instance.
+///
+/// This is used in combination with [FlexibleSyncConfiguration] to
+/// declare the set of queries you want to synchronize with the server. You can access and
+/// read the subscription set freely, but mutating it must happen in an [update]
+/// block.
+///
+/// Any changes to the subscription set will be persisted locally and be available the next
+/// time the application starts up - i.e. it's not necessary to subscribe for the same query
+/// every time. Updating the subscription set can be done while offline, and only the latest
+/// update will be sent to the server whenever connectivity is restored.
+///
+/// It is strongly recommended that you batch updates as much as possible and request the
+/// dataset your application needs upfront. Updating the set of active subscriptions for a
+/// Realm is an expensive operation serverside, even if there's very little data that needs
+/// downloading.
+/// {@category Sync}
+abstract class SubscriptionSet with IterableMixin<Subscription> {
+  Realm _realm;
+  SubscriptionSetHandle _handle;
+
+  SubscriptionSet._(this._realm, this._handle);
+
+  /// Finds an existing [Subscription] in this set by its query
+  ///
+  /// The [query] is represented by the corresponding [RealmResults] object.
+  /// Finds a subscription by query.
+  ///
+  /// Returns null, if not found
+  Subscription? find<T extends RealmObject>(RealmResults<T> query) {
+    final result = realmCore.findSubscriptionByResults(this, query);
+    return result == null ? null : Subscription._(result);
+  }
+
+  /// Finds an existing [Subscription] in this set by name.
+  ///
+  /// Returns null, if not found
+  Subscription? findByName(String name) {
+    final result = realmCore.findSubscriptionByName(this, name);
+    return result == null ? null : Subscription._(result);
+  }
+
+  Future<SubscriptionSetState> _waitForStateChange(SubscriptionSetState state) async {
+    final result = await realmCore.waitForSubscriptionSetStateChange(this, state);
+    realmCore.refreshSubscriptionSet(this);
+    return result;
+  }
+
+  /// Waits for the server to acknowledge the subscription set and return the matching objects.
+  ///
+  /// If the [state] of the subscription set is [SubscriptionSetState.complete]
+  /// the returned [Future] will complete immediately. If the state is
+  /// [SubscriptionSetState.error], the returned future will throw an
+  /// error.
+  Future<void> waitForSynchronization() async {
+    final result = await _waitForStateChange(SubscriptionSetState.complete);
+    if (result == SubscriptionSetState.error) {
+      throw RealmException('Synchronization Failed');
```

Comment:
Should this throw the actual error the server sent rather than this?

Reply:
There are no last error to fetch here - at least with the current c-api.